lib/Core/SQLiteBuildDB.cpp

```diff
@@ -373,32 +419,111 @@ class SQLiteBuildDB : public BuildDB {
     "DELETE FROM rule_dependencies WHERE rule_id == ?;";
   sqlite3_stmt* deleteFromRuleDependenciesStmt = nullptr;
 
-  virtual void setRuleResult(const Rule& rule,
-                             const Result& ruleResult) override {
+  //Return the key for a given keyID, keyID should exist in database.
+  KeyType keyForKeyID(uint64_t keyID) {
     int result;
-
-    // Find the existing rule id, if present.
-    //
-    // We rely on SQLite3 not using 0 as a valid rule ID.
-    result = sqlite3_reset(findIDForKeyInRuleResultsStmt);
+    result = sqlite3_reset(findKeyForKeyIDStmt);
+    assert(result == SQLITE_OK);
+    result = sqlite3_clear_bindings(findKeyForKeyIDStmt);
+    assert(result == SQLITE_OK);
+    result = sqlite3_bind_int64(findKeyForKeyIDStmt, /*index=*/1,
+                               keyID);
     assert(result == SQLITE_OK);
-    result = sqlite3_clear_bindings(findIDForKeyInRuleResultsStmt);
+    
+    KeyType key{};
+    
+    result = sqlite3_step(findKeyForKeyIDStmt);
+    if (result == SQLITE_DONE) {
+      abort();
+    } else if (result == SQLITE_ROW) {
+      assert(sqlite3_column_count(findKeyForKeyIDStmt) == 1);
+      return std::string(
+                (const char*)sqlite3_column_text(findKeyForKeyIDStmt, 0),
+                sqlite3_column_bytes(findKeyForKeyIDStmt, 0));
+    }
+    abort();
+  }
+  
+  //Finds keyID for a key and puts result in keyID_out
+  //returns false if key not present
+  bool findKeyIDForKey(const KeyType& key, uint64_t* keyID_out) {
+    int result;
+    result = sqlite3_reset(findKeyIDForKeyStmt);
+    assert(result == SQLITE_OK);
+    result = sqlite3_clear_bindings(findKeyIDForKeyStmt);
     assert(result == SQLITE_OK);
-    result = sqlite3_bind_text(findIDForKeyInRuleResultsStmt, /*index=*/1,
-                               rule.key.data(), rule.key.size(),
+    result = sqlite3_bind_text(findKeyIDForKeyStmt, /*index=*/1,
+                               key.data(), key.size(),
                                SQLITE_STATIC);
     assert(result == SQLITE_OK);
-
-    uint64_t ruleID = 0;
-    result = sqlite3_step(findIDForKeyInRuleResultsStmt);
+    
+    *keyID_out = 0;
+    result = sqlite3_step(findKeyIDForKeyStmt);
     if (result == SQLITE_DONE) {
-      ruleID = 0;
+      *keyID_out = 0;
+      return false;
     } else if (result == SQLITE_ROW) {
-      assert(sqlite3_column_count(findIDForKeyInRuleResultsStmt) == 1);
-      ruleID = sqlite3_column_int64(findIDForKeyInRuleResultsStmt, 0);
+      assert(sqlite3_column_count(findKeyIDForKeyStmt) == 1);
+      *keyID_out = sqlite3_column_int64(findKeyIDForKeyStmt, 0);
+      return true;
     } else {
       abort();
     }
+    
+    return false;
+  }
+  
+  //Adds a key into keys table if not present
+  //returns the keyID of the key
+  uint64_t addKey(const KeyType& key) {
```

Comment:
One overall general comment:

I think we should just have a single method for interacting with keys, which is to `getOrInsert(KeyType)` and it always returns an ID.

This means that we sometimes will end up creating a key even when `lookupRuleResult` will just realize the rule doesn't exist, but in practice the engine is just going to come along in a second and create that rule. This should simplify the patch and make only one thing that needs to be optimized.